### ISSUE TITLE
New skaffold version not compatible with docs/tutorials/tfx/template.ipynb

### ISSUE THREAD

**Issue by entrpn** [tensorflow/tfx#3191]
## URL(s) with the issue:

https://github.com/tensorflow/tfx/blob/master/docs/tutorials/tfx/template.ipynb

https://www.tensorflow.org/tfx/tutorials/tfx/template

## Description of issue (what needs changing):

Latest skaffold version breaks  pipeline creation in template.ipynb 

### Clear description:

Running through the  Kubeflow pipelines template  example on GCP and after installing the latest skaffold version as follows:

```bash
!curl -Lo skaffold curl -Lo skaffold curl -Lo skaffold https://storage.googleapis.com/skaffold/releases/v1.19.0/skaffold-linux-amd64 && chmod +x skaffold && mv skaffold /home/jupyter/.local/bin/
```

And then  running the pipeline creation:

```python
!tfx pipeline create  \
--pipeline-path=kubeflow_dag_runner.py \
--endpoint={ENDPOINT} \
--build-target-image={CUSTOM_TFX_IMAGE}
```

I get the following error:

```text
2021-02-03 19:33:27.260511: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/lib64:/usr/local/nccl2/lib:/usr/local/cuda/extras/CUPTI/lib64
2021-02-03 19:33:27.260578: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
CLI
Creating pipeline
Detected Kubeflow.
Use --engine flag if you intend to use a different orchestrator.
Reading build spec from build.yaml
Target image gcr.io/throw-away-294404/tfx-pipeline is not used. If the build spec is provided, update the target image in the build spec file build.yaml.
Use skaffold to build the container image.
/home/jupyter/.local/bin/skaffold
Error response from daemon: invalid reference format. Docker build ran into internal error. Please retry.
If this keeps happening, please open an issue..
No container image is built.
Traceback (most recent call last):
  File "/opt/conda/bin/tfx", line 10, in <module>
    sys.exit(cli_group())
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/click/decorators.py", line 73, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/tfx/tools/cli/commands/pipeline.py", line 117, in create_pipeline
    handler_factory.create_handler(ctx.flags_dict).create_pipeline()
  File "/opt/conda/lib/python3.7/site-packages/tfx/tools/cli/handler/kubeflow_handler.py", line 75, in create_pipeline
    skaffold_cmd)
  File "/opt/conda/lib/python3.7/site-packages/tfx/tools/cli/handler/kubeflow_handler.py", line 291, in _build_pipeline_image
    skaffold_cmd=skaffold_cmd).build()
  File "/opt/conda/lib/python3.7/site-packages/tfx/tools/cli/container_builder/builder.py", line 95, in build
    image_sha = skaffold_cli.build(buildspec_filename=self._buildspec.filename)
  File "/opt/conda/lib/python3.7/site-packages/tfx/tools/cli/container_builder/skaffold_cli.py", line 52, in build
    buildspec_filename
  File "/opt/conda/lib/python3.7/subprocess.py", line 411, in check_output
    **kwargs).stdout
  File "/opt/conda/lib/python3.7/subprocess.py", line 512, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['skaffold', 'build', '-q', '--output={{json .}}', '-f', 'build.yaml']' returned non-zero exit status 1.
```

The solution is to install skaffold v1.17.2 as follows:

```bash
!curl -Lo skaffold curl -Lo skaffold curl -Lo skaffold https://storage.googleapis.com/skaffold/releases/v1.17.2/skaffold-linux-amd64 && chmod +x skaffold && mv skaffold /home/jupyter/.local/bin/
```

**Comment by arghyaganguly** [comment]
@entrpn , I notice that the latest version for GoogleContainerTools/skaffold is v1.19.0 in link :-  https://github.com/GoogleContainerTools/skaffold/releases.

In  tensorflow.org/tfx/tutorials/tfx/cloud-ai-platform-pipelines, for 

#Download skaffold and set it executable, the command is 

!curl -Lo skaffold curl -Lo skaffold curl -Lo skaffold https://storage.googleapis.com/skaffold/releases/latest/skaffold-linux-amd64 && chmod +x skaffold && mv skaffold /home/jupyter/.local/bin/

Please confirm that similar issue persists for the above command(releases/latest).

**Comment by jiyongjung0** [comment]
@entrpn Thank you for reporting this.

By the way, the logs seems like not from the latest version (0.26.x) of TFX. We updated the TFX version of the notebook about a week ago. Could you let us know the TFX version you are using?

I've tried with skaffold 1.19, but I cound't reproduce your issue.

**Comment by entrpn** [comment]
Hello, thanks for  the quick response. Tried the following:

```bash
# Install tfx and kfp Python packages.
import sys
!{sys.executable} -m pip install --user --upgrade -q tfx==0.26.1
!{sys.executable} -m pip install --user --upgrade -q kfp==1.0.0
# Download skaffold and set it executable.
!curl -Lo skaffold curl -Lo skaffold curl -Lo skaffold https://storage.googleapis.com/skaffold/releases/latest/skaffold-linux-amd64 && chmod +x skaffold && mv skaffold /home/jupyter/.local/bin/
```

Then I check version:

```bash
!skaffold version
!python3 -c "from tfx import version ; print('TFX version: {}'.format(version.__version__))"
```

Output:

```text
v1.19.0
TFX version: 0.26.1
```

Then I run the create command again:

```bash
!tfx pipeline create  \
--pipeline-path=kubeflow_dag_runner.py \
--endpoint={ENDPOINT} \
--build-target-image={CUSTOM_TFX_IMAGE}
```

Same results:

```text
2021-02-05 21:20:45.460108: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/lib64:/usr/local/nccl2/lib:/usr/local/cuda/extras/CUPTI/lib64
2021-02-05 21:20:45.460160: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
CLI
Creating pipeline
Detected Kubeflow.
Use --engine flag if you intend to use a different orchestrator.
Reading build spec from build.yaml
Target image gcr.io/throw-away-294404/tfx-pipeline is not used. If the build spec is provided, update the target image in the build spec file build.yaml.
[Skaffold] Generating tags...
[Skaffold]  - gcr.io/throw-away-294404/tfx-pipeline -> gcr.io/throw-away-294404/tfx-pipeline:gcr.io/throw-away-294404/tfx-pipeline:latest
[Skaffold] Checking cache...
[Skaffold]  - gcr.io/throw-away-294404/tfx-pipeline: Not found. Building
[Skaffold] Building [gcr.io/throw-away-294404/tfx-pipeline]...
[Skaffold] Error response from daemon: invalid reference format. Docker build ran into internal error. Please retry.
[Skaffold] If this keeps happening, please open an issue..
No container image is built.
Traceback (most recent call last):
  File "/opt/conda/bin/tfx", line 8, in <module>
    sys.exit(cli_group())
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/click/decorators.py", line 73, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/jupyter/.local/lib/python3.7/site-packages/tfx/tools/cli/commands/pipeline.py", line 117, in create_pipeline
    handler_factory.create_handler(ctx.flags_dict).create_pipeline()
  File "/home/jupyter/.local/lib/python3.7/site-packages/tfx/tools/cli/handler/kubeflow_handler.py", line 75, in create_pipeline
    skaffold_cmd)
  File "/home/jupyter/.local/lib/python3.7/site-packages/tfx/tools/cli/handler/kubeflow_handler.py", line 291, in _build_pipeline_image
    skaffold_cmd=skaffold_cmd).build()
  File "/home/jupyter/.local/lib/python3.7/site-packages/tfx/tools/cli/container_builder/builder.py", line 92, in build
    image_sha = skaffold_cli.build(self._buildspec)
  File "/home/jupyter/.local/lib/python3.7/site-packages/tfx/tools/cli/container_builder/skaffold_cli.py", line 61, in build
    spec.filename))
RuntimeError: skaffold failed to build an image with build.yaml.
```

Then I revert back to  previous skaffold version:

```bash
# Install tfx and kfp Python packages.
import sys
!{sys.executable} -m pip install --user --upgrade -q tfx==0.26.1
!{sys.executable} -m pip install --user --upgrade -q kfp==1.0.0
# Download skaffold and set it executable.
!curl -Lo skaffold curl -Lo skaffold curl -Lo skaffold https://storage.googleapis.com/skaffold/releases/v1.17.2/skaffold-linux-amd64 && chmod +x skaffold && mv skaffold /home/jupyter/.local/bin/
#!curl -Lo skaffold curl -Lo skaffold curl -Lo skaffold https://storage.googleapis.com/skaffold/releases/latest/skaffold-linux-amd64 && chmod +x skaffold && mv skaffold /home/jupyter/.local/bin/
```

Run it again

```bash
!tfx pipeline create  \
--pipeline-path=kubeflow_dag_runner.py \
--endpoint={ENDPOINT} \
--build-target-image={CUSTOM_TFX_IMAGE}
```

Success:

```text
2021-02-05 21:25:50.822126: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/lib64:/usr/local/nccl2/lib:/usr/local/cuda/extras/CUPTI/lib64
2021-02-05 21:25:50.822177: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
CLI
Creating pipeline
Detected Kubeflow.
Use --engine flag if you intend to use a different orchestrator.
Reading build spec from build.yaml
Target image gcr.io/throw-away-294404/tfx-pipeline is not used. If the build spec is provided, update the target image in the build spec file build.yaml.
[Skaffold] Generating tags...
[Skaffold]  - gcr.io/throw-away-294404/tfx-pipeline -> time="2021-02-05T21:25:54Z" level=warning msg="{{.IMAGE_NAME}} is deprecated, envTemplate's template should only specify the tag value. See https://skaffold.dev/docs/pipeline-stages/taggers/"
[Skaffold] gcr.io/throw-away-294404/tfx-pipeline:latest
[Skaffold] Checking cache...
[Skaffold]  - gcr.io/throw-away-294404/tfx-pipeline: Not found. Building
[Skaffold] Building [gcr.io/throw-away-294404/tfx-pipeline]...
Sending build context to Docker daemon  1.317MBon    556kB
[Skaffold] Step 1/4 : FROM tensorflow/tfx:0.21.5
[Skaffold]  ---> 08a8ce5c97df
[Skaffold] Step 2/4 : WORKDIR /pipeline
[Skaffold]  ---> Using cache
[Skaffold]  ---> a1058d2a75d3
[Skaffold] Step 3/4 : COPY ./ ./
[Skaffold]  ---> 05eb4dd0b524
[Skaffold] Step 4/4 : ENV PYTHONPATH="/pipeline:${PYTHONPATH}"
[Skaffold]  ---> Running in e1abef45369c
[Skaffold]  ---> 6dfeb92f8d85
[Skaffold] Successfully built 6dfeb92f8d85
[Skaffold] Successfully tagged gcr.io/throw-away-294404/tfx-pipeline:latest
[Skaffold] The push refers to repository [gcr.io/throw-away-294404/tfx-pipeline]
[Skaffold] edce03f5205f: Preparing
[Skaffold] 4009e553a9ad: Preparing
[Skaffold] 5e54b7a48408: Preparing
[Skaffold] b41864cc7118: Preparing
[Skaffold] 918a92b87d0e: Preparing
[Skaffold] cc88d9e5114a: Preparing
[Skaffold] 6c30dbc09f4d: Preparing
[Skaffold] ff0d2decad59: Preparing
[Skaffold] 1c8005802d79: Preparing
[Skaffold] d62f92f20033: Preparing
[Skaffold] 84ff92691f90: Preparing
[Skaffold] 54b00d861a7a: Preparing
[Skaffold] c547358928ab: Preparing
[Skaffold] 84ff92691f90: Preparing
[Skaffold] c4e66be694ce: Preparing
[Skaffold] 47cc65c6dd57: Preparing
[Skaffold] cc88d9e5114a: Waiting
[Skaffold] 6c30dbc09f4d: Waiting
[Skaffold] ff0d2decad59: Waiting
[Skaffold] 1c8005802d79: Waiting
[Skaffold] d62f92f20033: Waiting
[Skaffold] 84ff92691f90: Waiting
[Skaffold] 54b00d861a7a: Waiting
[Skaffold] c547358928ab: Waiting
[Skaffold] c4e66be694ce: Waiting
[Skaffold] 47cc65c6dd57: Waiting
[Skaffold] 918a92b87d0e: Layer already exists
[Skaffold] 4009e553a9ad: Layer already exists
[Skaffold] 5e54b7a48408: Layer already exists
[Skaffold] b41864cc7118: Layer already exists
[Skaffold] 1c8005802d79: Layer already exists
[Skaffold] ff0d2decad59: Layer already exists
[Skaffold] cc88d9e5114a: Layer already exists
[Skaffold] 6c30dbc09f4d: Layer already exists
[Skaffold] d62f92f20033: Layer already exists
[Skaffold] 84ff92691f90: Layer already exists
[Skaffold] 54b00d861a7a: Layer already exists
[Skaffold] c547358928ab: Layer already exists
[Skaffold] c4e66be694ce: Layer already exists
[Skaffold] 47cc65c6dd57: Layer already exists
[Skaffold] edce03f5205f: Pushed
[Skaffold] latest: digest: sha256:21019158eb0cbc93026a27d308a116ad1051cff71876078e04c8ac78b610863c size: 3693
[Skaffold] There is a new version (1.19.0) of Skaffold available. Download it from:
[Skaffold]   https://github.com/GoogleContainerTools/skaffold/releases/tag/v1.19.0
[Skaffold] 
[Skaffold] Help improve Skaffold with our 2-minute anonymous survey: run 'skaffold survey'
New container image is built. Target image is available in the build spec file.
2021-02-05 21:26:06.438547: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/lib64:/usr/local/nccl2/lib:/usr/local/cuda/extras/CUPTI/lib64
2021-02-05 21:26:06.438607: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
Traceback (most recent call last):
  File "kubeflow_dag_runner.py", line 24, in <module>
    from pipeline import pipeline
  File "/home/jupyter/imported/my_pipeline/pipeline/pipeline.py", line 28, in <module>
    from tfx.components import BigQueryExampleGen  # pylint: disable=unused-import
ImportError: cannot import name 'BigQueryExampleGen' from 'tfx.components' (/home/jupyter/.local/lib/python3.7/site-packages/tfx/components/__init__.py)
Error while running "/opt/conda/bin/python kubeflow_dag_runner.py" 
```

**Comment by jiyongjung0** [comment]
Thank you so much for the detailed description. The relevant error messsage is

"[Skaffold] Error response from daemon: invalid reference format. Docker build ran into internal error. Please retry.".

It seems like your `build.yaml` file is not compatible with latest version of skaffold. Could you share your build.yaml file, too?
Because this file is not automatically updated, it might be generated from older versions of TFX. (Although it should be compatible with recent TFX/Skaffold...) It seems like there is missing problem....

**Comment by entrpn** [comment]
Here you go.

```yaml
apiVersion: skaffold/v1beta13
build:
  artifacts:
  - context: .
    docker:
      dockerfile: Dockerfile
    image: gcr.io/throw-away-294404/tfx-pipeline
  tagPolicy:
    envTemplate:
      template: '{{.IMAGE_NAME}}:latest'
kind: Config
```

Thank you.

**Comment by jiyongjung0** [comment]
**tl;dr**: change `'{{.IMAGE_NAME}}:latest'` to `latest`.

Oh, now all pieces matched together.

Skaffold has marked '{{.IMAGE_NAME}}' in template as deprecated several months ago.[Related skaffold doc](https://skaffold.dev/docs/pipeline-stages/taggers/#envtemplate-uses-values-of-environment-variables-as-tags)

TFX also changed implementation not to use '{{.IMAGE_NAME}}' from 0.23.0[commit](https://github.com/tensorflow/tfx/commit/7ffd56bb96569c26be91fe3d05ca3327c01e7106#diff-b424fcaaadc7a796b15b54b4c20946a705899afe37438449275d43965abd1f24). So the build.yaml file created with 0.23.0 or later doesn't have that part.
The change is simply replacing `'{{.IMAGE_NAME}}:latest'` to `latest`. (No quote needed)

What I didn't aware of was that skaffold really stopped supporting IMAGE_NAME because it worked for a while although it is deprecated. But it seems they dropped support of deprecated feature in recent versions.

Thank you for sharing this information and providing a lot of clues for this chase.

**Comment by entrpn** [comment]
Yup that was exactly the issue. Thank you.

A follow up question, I believe  the build.yaml gets generated when I run:

```
!tfx pipeline create  \
--pipeline-path=kubeflow_runner.py \
--endpoint={ENDPOINT} \
--build-target-image={CUSTOM_TFX_IMAGE}
```
So I'm wondering if something somewhere needs updating as well?

Again, thank you for your prompt replies.

**Comment by jiyongjung0** [comment]
Thank you for the confirmation!

Using the updated TFX version should be enough. Currently we don't update an existing `build.yaml` file, and just use it to build a container image. I think that TFX CLI should become more robust on this kind of changes on dependencies. For example,
- updating `build.yaml` and `Dockerfile` automatically.
- don't depend on skaffold and use docker API directly.

But we don't have concrete plan yet, and it would be great if you can give any suggestions or feedbacks.
Thanks.

**Comment by entrpn** [comment]
As of right now, I don't know enough to make suggestions. As I get more familiar, I'll be sure to do so.

Thank you for all your help. I'm closing this.
